src/Microsoft.ML.Api/PredictionEngine.cs

```diff
@@ -132,20 +155,22 @@ public void Reset()
     /// </summary>
     /// <typeparam name="TSrc">The user-defined type that holds the example.</typeparam>
     /// <typeparam name="TDst">The user-defined type that holds the prediction.</typeparam>
-    public sealed class PredictionEngine<TSrc, TDst>
+    public abstract class PredictionEngineBase<TSrc, TDst>
         where TSrc : class
         where TDst : class, new()
     {
         private readonly DataViewConstructionUtils.InputRow<TSrc> _inputRow;
         private readonly IRowReadableAs<TDst> _outputRow;
         private readonly Action _disposer;
 
-        internal PredictionEngine(IHostEnvironment env, Stream modelStream, bool ignoreMissingColumns,
+        [BestFriend]
+        internal PredictionEngineBase(IHostEnvironment env, Stream modelStream, bool ignoreMissingColumns,
```

Comment:
Incidentally, given that this is a base class, there is a new concept in either C# 7.2 or 7.3 or something called "private protected" which is precisely for this scenario, perhaps you could use it here, now that this is a base class.